File: android/src/main/java/to/holepunch/bare/android/HostPluginRegistry.kt
package to.holepunch.bare.android

import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

/**
 * Registers host-side invoke handlers for events delegated from the Bare worklet over IPC.
 * Handlers respond asynchronously ([respond] may be called later, e.g. after an Activity
 * result), mirroring how native pickers/cameras return their payloads.
 */
class HostPluginRegistry {
    data class Key(val pluginId: String, val event: String)

    fun interface Handler {
        fun invoke(args: JSONObject?, payload: ByteArray?, respond: (HostInvokeOutcome) -> Unit)
    }

    sealed class HostInvokeOutcome {
        data class Ok(val value: JSONObject) : HostInvokeOutcome()
        data class Fail(val code: String, val message: String) : HostInvokeOutcome()
    }

    // ConcurrentHashMap: register() runs on the host app main thread while
    // dispatch()/toCapabilitiesJson() run on Bare Kit IPC callback threads, so
    // the shared map must be safe for concurrent mutation + iteration. Iteration
    // is weakly consistent (never throws ConcurrentModificationException) and
    // toCapabilitiesJson snapshots into a private map for a consistent view.
    private val handlers = ConcurrentHashMap<Key, Handler>()

    fun register(pluginId: String, event: String, handler: Handler) {
        handlers[Key(pluginId, event)] = handler
    }

    fun toCapabilitiesJson(): JSONArray {
        val snapshot = LinkedHashMap(handlers)
        val byPlugin = mutableMapOf<String, MutableSet<String>>()
        for ((key, _) in snapshot) {
            byPlugin.getOrPut(key.pluginId) { mutableSetOf() }.add(key.event)
        }
        val arr = JSONArray()
        for ((pluginId, events) in byPlugin) {
            val evArr = JSONArray()
            for (e in events.sorted()) evArr.put(e)
            arr.put(
                JSONObject().apply {
                    put("pluginId", pluginId)
                    put("capabilities", JSONArray())
                    put("events", evArr)
                    put("runtimes", JSONArray().put("android"))
                },
            )
        }
        return arr
    }

    fun dispatch(
        pluginId: String,
        event: String,
        args: JSONObject?,
        payload: ByteArray?,
        respond: (HostInvokeOutcome) -> Unit,
    ) {
        val h = handlers[Key(pluginId, event)]
            ?: return respond(
                HostInvokeOutcome.Fail(
                    ErrorCodes.UNSUPPORTED_CAPABILITY,
                    "No host handler for $pluginId.$event",
                ),
            )
        h.invoke(args, payload, respond)
    }
}

File: ios/Sources/BareHost/HostPluginRegistry.swift
import Foundation

/**
 * Registers host-side invoke handlers for events delegated from the Bare
 * worklet over IPC. Mirror of `android/.../HostPluginRegistry.kt`.
 */
public final class HostPluginRegistry {
  public struct Key: Hashable {
    public let pluginId: String
    public let event: String

    public init(pluginId: String, event: String) {
      self.pluginId = pluginId
      self.event = event
    }
  }

  public enum HostInvokeOutcome {
    case ok([String: Any])
    case fail(code: String, message: String)
  }

  /// Handlers respond asynchronously ([respond] may be called later, e.g. after
  /// a native picker/camera result), mirroring the Kotlin registry.
  public typealias Handler =
    ([String: Any]?, Data?, @escaping (HostInvokeOutcome) -> Void) -> Void

  /// `handlers` is mutated from `register` (main/UI thread) and read from
  /// `dispatch` (an IPC Task) — guard with an unfair lock to avoid a data race.
  private var handlers: [Key: Handler] = [:]
  private var lock = os_unfair_lock()

  public init() {}

  public func register(
    pluginId: String,
    event: String,
    handler: @escaping Handler
  ) {
    os_unfair_lock_lock(&lock)
    handlers[Key(pluginId: pluginId, event: event)] = handler
    os_unfair_lock_unlock(&lock)
  }

  /** Capability rows for `HOST_CAPABILITIES_RESPONSE`, runtimes `["ios"]`. */
  public func toCapabilitiesJSON() -> [[String: Any]] {
    os_unfair_lock_lock(&lock)
    defer { os_unfair_lock_unlock(&lock) }
    var byPlugin: [String: Set<String>] = [:]
    for key in handlers.keys {
      var events = byPlugin[key.pluginId] ?? []
      events.insert(key.event)
      byPlugin[key.pluginId] = events
    }
    return byPlugin.keys.sorted().map { pluginId in
      let events = (byPlugin[pluginId] ?? []).sorted()
      return [
        "pluginId": pluginId,
        "capabilities": [],
        "events": events,
        "runtimes": ["ios"],
      ]
    }
  }

  public func dispatch(
    pluginId: String,
    event: String,
    args: [String: Any]?,
    payload: Data?,
    respond: @escaping (HostInvokeOutcome) -> Void
  ) {
    os_unfair_lock_lock(&lock)
    let handler = handlers[Key(pluginId: pluginId, event: event)]
    os_unfair_lock_unlock(&lock)
    // Fixed string only — never reflect caller-controlled pluginId/event back
    // to page JS (would enable capability probing).
    guard let handler else {
      respond(
        .fail(
          code: ErrorCodes.unsupportedCapability,
          message: "No host handler for the requested capability"
        )
      )
      return
    }
    handler(args, payload, respond)
  }
}
